bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java

```
@@ -0,0 +1,115 @@
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+@Component(label = "Demandware Demandware Client Provider", immediate = true)
+@Service(value = DemandwareClientProvider.class)
+public class DemandwareClientProviderImpl implements DemandwareClientProvider {
+	private static final Logger LOG = LoggerFactory.getLogger(DemandwareClientProviderImpl.class);
+	private static final String DWRE_SCHEME = "demandware://";
+	public static final String INSTANCE_ID_DEFAULT = "default";
+	@Reference(referenceInterface = DemandwareClient.class,
+			bind = "bindDemandwareClient",
+			unbind = "unbindDemandwareClient",
+			cardinality = ReferenceCardinality.MANDATORY_MULTIPLE, // TODO check if works with one
+			policy = ReferencePolicy.DYNAMIC)
+	protected HashMap<String, DemandwareClient> demandwareClients;
+
+	/**
+	 * Returns DemandwareClient service, configured with "default" instance id
+	 * or, if not present, first found DemandwareClient service.
+	 *
+	 * @return
+	 *///TODO remove usages
+	public DemandwareClient getDefaultClient() {
+		Optional<DemandwareClient> client = getClientForSpecificInstance(INSTANCE_ID_DEFAULT);
+		if (client.isPresent()) {
+			return client.get();
+		}
+
+		LOG.debug("No default DemandwareClient configured: return first found service.");
+		Optional<DemandwareClient> clientOpt = demandwareClients.entrySet()
+				.stream()
+				.findFirst()
+				.map(Map.Entry::getValue);
+		if (clientOpt.isPresent()) {
+			return clientOpt.get();
+		}
+
+		LOG.error("Failed to get DemandwareClient - no configuration found.");
+		return null;
+	}
+
+	@Override
+	public Optional<DemandwareClient> getClientForSpecificInstance(final String instanceId) {
+		DemandwareClient demandwareClient = demandwareClients.get(instanceId.replace("/", ""));
+		if (demandwareClient == null) {
+			LOG.error("DemandwareClient not found for instanceId [{}]", instanceId);
+			return Optional.empty();
+		}
+		return Optional.of(demandwareClient);
```

this is not "optional" when you can't provide a client, you cannot continue in the upper processes. we should throw an exception. not having a service is a serious problem. btw: when you throw excpetions, don't log.